[Feature] Distributed Training for Trainer

mlahariya opened the pull request:
Here is the distributed training setup for the Trainer. This MR aims to solve #307, and #670

![image](https://github.com/user-attachments/assets/8219b0b0-c53f-47c4-82b9-ac87aca6bcc1)

We define an Accelerator class, that takes care of distribution of tasks. A Distributor class takes care of setting up and managing all the processes. The Execution takes care of each processes execution setup (such as environment variables, device, dtype, etc)
There features can be used using the nporcs>1 method in train config. 

Tasks
   - [x] Set the Distribution
   - [x] Set up Execution
   - [x] Set Up the accelerator
   - [x] Integrate Accelerator in Trainer - wrapper for fit()
   - [x] Update Callbacks to only work with process 0
   - [x] Update TrainConfig with new Parameters
   - [x] Prepare model, opt, dataloader, and batches in trainer using accelerator
   - [x] Optimize space for issue #670

mlahariya commented:
Hey @RolandMacDoland , @chMoussa , @debrevitatevitae - This is ready for review. Thanks!

chMoussa commented on qadence/ml_tools/config.py:
```diff
@@ -203,6 +204,67 @@ def validation_criterion(val_loss: float, best_val_loss: float, val_epsilon: flo
+
+    For effective parallel processing, set this to a value greater than 1.
+    - In case of Multi-GPU or Multi-Node-Multi-GPU setups, nprocs should be equal to
+    the total world_size, or total number of GPU to be used.
```
What is `world_size`?

RolandMacDoland replied:
I guess this is the number of GPUs that the whole application requires.

mlahariya replied:
Indeed. its the total number of GPUs. I will update it so that its less confusing.

chMoussa commented on qadence/ml_tools/config.py:
```diff
@@ -203,6 +204,67 @@ def validation_criterion(val_loss: float, best_val_loss: float, val_epsilon: flo
+    If None, the default (usually torch.float32) is used.
+    """
+
+    all_reduce_metrics: bool = False
```
So if I understand correctly, in background, this is equivalent to concatenating arrays predictions/labels, to then call the metrics functional, correct? Or the user has to specify somewhere else how the aggregation takes place?

mlahariya replied:
This is just a functionality provided to the user. Its equivalent to averaging all loss & metrics across all processes - other wise only the main process loss & metrics will be used. 

I dont think we need to provide the functionality of how the aggregation takes place (average/sum/etc) - because generally speaking - one averages a value across processes to get its unbiased estimate. It is also usually avoided, because we are mostly interested in the main process.

chMoussa commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+        For now it only supports SLURM Cluster, and should be extended to others
+        when needed.
+        """
+        self.job_id = int(os.environ.get("SLURM_JOB_ID", 93345))
```
Why 93345 as default?

mlahariya replied:
Same as above.

chMoussa commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+        For now it only supports SLURM Cluster, and should be extended to others
+        when needed.
+        """
+        self.job_id = int(os.environ.get("SLURM_JOB_ID", 93345))
```
Why 93345 as default?

mlahariya replied:
Just a random job ID I was testing it on. Can also change to a random.choice - but that might lead to issues in identifying ports - as unavailable ports might get selected.

mlahariya replied:
self.job_id is an indicator used to assign id to the execution (and to create a unique port every run). It has nothing to do with slurm. It should be assigned for every run, on local run/slurm cluster/any other cluster. 

SLURM_JOB_ID is being used as a proxy for this self.job_id - as its always unique - and we wont face port unavailable issues on the cluster. 


On your local run - the localhost has port 21069 (based on default job id 93345 ((93345 + 12000) / 5000))
On slurm cluster - the nodexx has port xxxxx (based on the slurm job id ((SLURM_JOB_ID + 12000) / 5000))

I don't think we should raise an error. What we can do is change the default value, which will also change the default port. We can either hard-code it, or, randomly generate the job_id (where we risk getting port unavailability.)

mlahariya replied:
Alright - I have updated it so that we decouple job_id and port. 
Port is identified using a _find_avaliable_port function. 
And job_id is "Unknown" as a default.

chMoussa commented on tests/ml_tools/test_accelerator.py:
```diff
@@ -0,0 +1,281 @@
+    return DataLoader(TensorDataset(x, y), batch_size=batch_size)
+
+
+@pytest.fixture
+def dummy_model() -> DummyModel:
+    return DummyModel()
+
+
+@pytest.fixture
+def dummy_optimizer(dummy_model: DummyModel) -> optim.Optimizer:
+    return get_dummy_optimizer(dummy_model)
+
+
+@pytest.fixture
+def dummy_dataloader() -> DataLoader:
+    return get_dummy_dataloader(batch_size=4)
+
+
+@pytest.fixture
+def dict_dataloader() -> DictDataLoader:
+    data_configs: Dict[str, Dict[str, int]] = {
+        "dataset1": {"data_size": 30, "batch_size": 5},
+        "dataset2": {"data_size": 50, "batch_size": 10},
+    }
+    dataloaders: Dict[str, DataLoader] = {}
+    for name, cfg in data_configs.items():
+        x = torch.randn(cfg["data_size"], 10)
+        y = torch.randn(cfg["data_size"], 1)
+        dataloaders[name] = to_dataloader(x, y, batch_size=cfg["batch_size"])
+    return DictDataLoader(dataloaders)
```
Shall we separate all ml_tools fixtures into a `conftest.py` file?

mlahariya replied:
This is a very good idea. We can create a issue about it - but I don't think we need to do it in this MR (which already is quite big)

chMoussa commented on tests/ml_tools/test_callbacks.py:
```diff
@@ -210,15 +210,17 @@ def test_early_stopping(trainer: Trainer) -> None:
     trainer.opt_result.metrics = {monitor_metric: 0.5}
     callback(stage, trainer, trainer.config, writer)
-    assert trainer.stop_training is False
+    assert trainer._stop_training.detach().item() == 0
```
So remind me again what's the motivation to change _stop_training to a torch Tensor?

mlahariya replied:
Because we need to all_reduce it during the training - and bool variables can't be reduced. 

i.e. we need to share the information across all processes if the training should be stopped (e.g. if process 1 triggers early stopping, all processes should stop - the _stop_training indicator from process 1 gets updated in process 1 - then we do all reduce to share this information across all processes, and all processes stop.)

the function stop_training in trainer returns bool.
the variable _stop_training is a torch tensor, which is shared across processes.

RolandMacDoland commented on docs/tutorials/qml/ml_tools/data_and_config.md:
```diff
@@ -206,12 +215,52 @@ config = TrainConfig(
 
+> Notes:
+> - *Logging Specific Callbacks*: Logging is availiable only through the main process, i.e. process 0.  Model logging, plotting, logging metrics will only be performed for a single process, even is multiple processes are run.
+> - *Training specific callbacks*: Callbacks specific to training, e.g., `EarlyStopping`, `LRSchedulerStepDecay`, etc will be called from each process.
```
```suggestion
> - *Training with specific callbacks*: Callbacks specific to training, e.g., `EarlyStopping`, `LRSchedulerStepDecay`, etc will be called from each process.
```

mlahariya replied:
Updated

RolandMacDoland commented on docs/tutorials/qml/ml_tools/data_and_config.md:
````diff
@@ -206,12 +215,52 @@ config = TrainConfig(
+)
+```
+
+Example: For GPU Training
````
```suggestion
Example: For GPU multiprocessing training
```

mlahariya replied:
Updated.

RolandMacDoland commented on docs/tutorials/qml/ml_tools/intro.md:
```diff
@@ -0,0 +1,34 @@
+# Introduction to Qadence ML Tools
+
+Welcome to the Qadence `ML Tools` documentation! This submodule is designed to streamline your machine learning workflows—especially for quantum machine learning—by providing a set of robust tools for training, monitoring, and optimizing your models.
```
```suggestion
Welcome to the Qadence `ML Tools` documentation. This submodule is designed to streamline your machine learning workflows —especially for quantum machine learning— by providing a set of robust tools for training, monitoring, and optimizing your models.
```

RolandMacDoland commented on docs/tutorials/qml/ml_tools/intro.md:
```diff
@@ -0,0 +1,34 @@
+
+Welcome to the Qadence `ML Tools` documentation! This submodule is designed to streamline your machine learning workflows—especially for quantum machine learning—by providing a set of robust tools for training, monitoring, and optimizing your models.
+
+## What You'll Find in This Documentation
```
```suggestion
## What this documentation is about
```

RolandMacDoland commented on docs/tutorials/qml/ml_tools/trainer.md:
```diff
@@ -398,15 +396,14 @@ config = TrainConfig(
 trainer = Trainer(model, optimizer, config, loss_fn=loss_fn,
-                    train_dataloader = train_dataloader, val_dataloader = val_dataloader,
-                    device=DEVICE, dtype=DTYPE)
+                    train_dataloader = train_dataloader, val_dataloader = val_dataloader)
```
```suggestion
                    train_dataloader=train_dataloader, val_dataloader=val_dataloader)
```

RolandMacDoland commented on qadence/ml_tools/callbacks/callback.py:
```diff
@@ -95,14 +95,36 @@ def __init__(
+            )
+
+    @staticmethod
+    def default_callback(_: Any) -> bool:
```
hm, what is this supposed to do ?

mlahariya replied:
This is used in the default route for otherwise custom specified callback condition. Carried on from the legacy code.

If user wants to define a custom callback condition - they can provide a callback_condition in the Callback init (e.g. if they want to run the callback at every prime numbered iteration - this function will return true for prime numbers)
The default is our setup. Which should bypass this, and always return true.

It used to be a lambda function - but process spawning does not allow lambda functions - because it needs all functions to pickle-able. 

Thank you for noticing this. It was a pain to identify this bug and resolve it. 
We can also completely remove the callback_condition argument to make things easier. Its not being used much (afaik).

RolandMacDoland commented on qadence/ml_tools/data.py:
```diff
@@ -34,6 +34,10 @@ class OptimizeResult:
     """Metrics that can be saved during training."""
     extra: dict = field(default_factory=lambda: dict())
     """Extra dict for saving anything else to be used in callbacks."""
+    rank: int | None = 0
```
hm not sure why `rank` could be of `None` type if you default it to 0.

mlahariya replied:
Updated. Removed None.

RolandMacDoland reviewed (commented):
Thanks @mlahariya for this impressive piece of work! For the sake of mental health, I'll break the review into chunks. This is the first bit and I'll get over from `accelerator.py` next,

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+        log_device (str | None): Logging device, e.g., "cpu" or "cuda:<local_rank>".
+        dtype (torch.dtype): Data type for controlling numerical precision (e.g., torch.float32).
+        data_dtype (torch.dtype): Data type for controlling datasets precision (e.g., torch.float16).
+    """
```
Attrs are missing, such as `spawn`, `nprocs`, `strategy`, `node_rank`

mlahariya replied:
Updated.

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+
+    def __init__(
+        self,
+        compute_setup: str = "auto",
```
Maybe rename to `compute_environment`

mlahariya replied:
I'm not sure about this. Lets keep it compute_setup for now to avoid confusion with other variables that contain environment in the names.

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+            self.data_dtype = torch.float64 if (self.dtype == torch.complex128) else torch.float32
+        self._set_cluster_variables()
+
+    def detect_strategy(self) -> str:
```
why not letting detect strategy change `self.strategy` directly (and call this method in `__init__`)?

mlahariya replied:
self.spawn is needed in detect strategy. Which is setup in the accelerator. we detect the strategy in the accelerators __init__.

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+            int(torch.cuda.device_count()) if torch.cuda.is_available() else 1
+        )
+
+    def setup_environment(self, process_rank: int) -> tuple[int, int, int | None]:
```
Maybe rename to `setup_distributed_environment`

mlahariya replied:
Updated.

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+    def setup_environment(self, process_rank: int) -> tuple[int, int, int | None]:
+        """
+        Set up environment variables and the computation device for distributed processing.
+
```
it would help here to know what all the different ranks mean

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+        """
+        # set the process based variables
+        self.node_rank = int(os.environ.get("SLURM_NODEID", 0))
+        self.node_name = os.environ.get("SLURMD_NODENAME", "Unknown")
```
Typo `SLURMD_`

mlahariya replied:
No. Thats a default slurm variable. I dont know why the use SLURMD_ :/ https://slurm.schedmd.com/slurm.conf.html#OPT_SLURMD_NODENAME

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+        os.environ["MASTER_ADDR"] = self.master_addr = self._get_master_addr()
+        os.environ["MASTER_PORT"] = self.master_port = self._get_master_port()
+
+        return self.rank, self.world_size, self.local_rank
```
Consider returning a dict for clarity.

mlahariya replied:
Updated.

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+        """
+        if "MASTER_ADDR" in os.environ:
+            return os.environ["MASTER_ADDR"]
+        elif self.strategy == "default":
```
[Refactoring] strategy is used all over to implement a default behaviour to environment setting. My suggestion: isolate a class `DefaultEnvironmentStrategy`, implement the same methods name as here and call those in the related methods of `DistributionStrategy`.

mlahariya replied:
I have updated according to this suggestion. 
1) We have an execution baseclass, which can be inherited to different types of executions 
 - we have added default and torchrun executions 
2) Added a Distributor class that uses the execution to properly distribue and manage processes.

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+    def _get_local_rank(self, process_rank: int) -> int | None:
+        """
+        Retrieve the local rank of the current process (its index on the local node).
+
```
Missing args

mlahariya replied:
Updated.

debrevitatevitae commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+            return int(process_rank)
+        return 0
+
+    def _set_compute(self) -> None:
```
Maybe rename to `_finalize_compute_environment`

mlahariya replied:
I am not sure about this. This method only sets the compute - to finalize the environment - we need to also set the device and local env variables.

debrevitatevitae reviewed (commented):
Hey @mlahariya , thanks for the huge work! I also didn’t have the chance to give a full review yet, but I focused on `strategy.py`. I think that some renaming and a bit of refactoring there, maybe with a strategy design pattern, would really help readability and maintainability.

mlahariya commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+    It also provides methods to start and clean up the PyTorch distributed process group.
+
+    Attributes:
+        backend (str): The backend used for distributed communication (e.g., "nccl", "gloo").
```
Updated

mlahariya commented on docs/tutorials/qml/ml_tools/data_and_config.md:
```diff
@@ -206,12 +215,52 @@ config = TrainConfig(
 
 
+> Notes:
+> - *Logging Specific Callbacks*: Logging is availiable only through the main process, i.e. process 0.  Model logging, plotting, logging metrics will only be performed for a single process, even is multiple processes are run.
```
Updated.

mlahariya commented on docs/tutorials/qml/ml_tools/CPU.md:
```diff
@@ -0,0 +1,102 @@
+### Understanding Arguments
+- *spawn*: If True, enables multi-processing and launches separate processes.
+- *nprocs*: Number of processes to run.
+- *compute_setup*: The computational setup used for training. Options include `cpu`, `gpu`, and `auto`.
```
Device is a default terminology for the local compute - ("cpu", "cuda:0", "cuda:1", etc). It would be convoluted to use this for the whole setup.

There is no standard terminology for the whole setup - some libraries use `accelerator` (https://lightning.ai/docs/pytorch/stable/accelerators/gpu_basic.html), some use `strategy`.

mlahariya commented on docs/tutorials/qml/ml_tools/CPU.md:
```diff
@@ -0,0 +1,102 @@
+
+### Single-Process Training Configuration:
+- **`backend="cpu"`**: Ensures training runs on the CPU.
+- **`spawn=False`**: Uses a single process (default).
+- **`nprocs=1`**: Uses one CPU core.
```
If spawn == False => nprocs will be ignored. 

`spawn` indicates if the processes should be spawned, and `nprocs` indicate the number of processes. I divided the two to keep flexibility - but we can also remove the spawn, and only use `nprocs`.

RolandMacDoland replied:
Yeah, that's what I've hinted at. As far as I recall, Slurm scripts don't require it so I assume it'll make the whole interface for user friendly with one arg less.

mlahariya commented on docs/tutorials/qml/ml_tools/CPU.md:
```diff
@@ -0,0 +1,102 @@
+
+### Multi-Processing Configuration
+- **`backend="gloo"`**: Uses the Gloo backend for CPU multi-processing.
+- **`spawn=True`**: Enables multi-processing.
+- **`nprocs=4`**: Utilizes 4 CPU cores.
```
I am not sure I understand this. The backend dictates how the communication happens between processes, and should be supported backend by `torch.distributed`.

mlahariya commented on docs/tutorials/qml/ml_tools/CPU.md:
```diff
@@ -0,0 +1,102 @@
+    # Multi-Process Training Configuration
+    train_config = TrainConfig(
+        compute_setup="cpu",
+        backend="gloo",
```
I added a bit more information about the backend. https://pytorch.org/docs/stable/distributed.html#torch.distributed.Backend

mlahariya commented on docs/tutorials/qml/ml_tools/GPU.md:
```diff
@@ -0,0 +1,267 @@
+This guide explains how to train models on **GPU** using `Trainer` from `qadence.ml_tools`, covering **single-GPU**, **multi-GPU (single node)**, and **multi-node multi-GPU** setups.
+
+### Understanding Arguments
+- *spawn*: If True, enables multi-processing and launches separate processes.
+- *nprocs*: Number of processes to run.
```
Open in another comment chain.

mlahariya commented on docs/tutorials/qml/ml_tools/GPU.md:
````diff
@@ -0,0 +1,267 @@
+- **`nprocs=1`**: Uses one GPU.
+```python
+train_config = TrainConfig(
+    compute_setup="auto",
````
default mode is `cpu`. `auto` is useful when running on GPU clusters, in case the GPUs are unavailable, it falls back to CPUs. otherwise if `gpu` is used, and GPUs are unavailable - we through an error.

mlahariya commented on docs/tutorials/qml/ml_tools/GPU.md:
```diff
@@ -0,0 +1,267 @@
+- **`compute_setup`**: Selected training setup. (`gpu` or `auto`)
+- **`backend="nccl"`**: Required for multi-node setups.
+- **`spawn=True`**: Enables multi-node processing.
+- **`nprocs=4`**: Uses 4 GPUs across nodes.
```
Here we are using 2 GPUs per node - and a total of 2 nodes. 
so we use a total of 4 GPUs. 
Which run 4 processes. 

`nprocs` is the total number of GPUs across all nodes.

mlahariya commented on docs/tutorials/qml/ml_tools/GPU.md:
```diff
@@ -0,0 +1,267 @@
+
+## Examples
+
+The following sections provide Python scripts and SLURM batch scripts for each setup.
```
Updated.

mlahariya commented on docs/tutorials/qml/ml_tools/GPU.md:
```diff
@@ -0,0 +1,267 @@
+Trainer.set_use_grad(True)
+
+# __main__ is recommended.
+if __name__ == "__main__":
```
These are just simple model and optimizer (which I randomly made to showcase how to use Trainer) - nothing fancy. I will add comments though.

mlahariya commented on docs/tutorials/qml/ml_tools/GPU.md:
```diff
@@ -0,0 +1,267 @@
+
+- *Assuming that you have 1 node1 with 2 GPU. These numbers can be changed depending on user needs.*
+
+You can train by sinply calling this on the head node.
```
Yes. Head node should be the standard terminology for the main node that orchestrates the processes across resources (?)

mlahariya commented on docs/tutorials/qml/ml_tools/GPU.md:
```diff
@@ -0,0 +1,267 @@
+srun torchrun \
+--nnodes 1 \
+--nproc_per_node 2 \
+--rdzv_id $RANDOM \
```
These are standard torchrun flags - to specify the process parameters on torchrun. https://pytorch.org/docs/stable/elastic/run.html

mlahariya commented on docs/tutorials/qml/ml_tools/GPU.md:
```diff
@@ -0,0 +1,267 @@
+--nproc_per_node 2 \
+--rdzv_id $RANDOM \
+--rdzv_backend c10d \
+--rdzv_endpoint $head_node_ip:29501 \
```
Not exactly. But its good practice. 

We need a head_node_ip for any one node in the cluster. This could be the first node, head node, or any other node. The only constraint is that same head_node_ip should be provided to all processes. I choose the first node as the head node - but you can also choose 2nd/3rd (which might cause issues in case you only have 1 node)


I will add this in the documentation.

mlahariya commented on docs/tutorials/qml/ml_tools/accelerator_doc.md:
```diff
@@ -0,0 +1,202 @@
+    nprocs=4,               # Number of processes (e.g., GPUs)
+    compute_setup="auto",   # Automatically selects available compute devices
+    log_setup="cpu",        # Logs on CPU to avoid memory overhead
+    dtype=torch.float32,    # Data type for numerical precision
```
No we don't need to set it up explicitly. 
If not provided, the torch default will be used. 
I just added this here - to showcase that we also have the functionality to specify dtype for model and data.

mlahariya commented on docs/tutorials/qml/ml_tools/accelerator_doc.md:
```diff
@@ -0,0 +1,202 @@
+
+            batch_data, batch_targets = batch
+            optimizer.zero_grad()
+            output = model(batch_data)
```
Indeed. We train the model - for each batch of the dataset.

mlahariya commented on docs/tutorials/qml/ml_tools/accelerator_doc.md:
```diff
@@ -0,0 +1,202 @@
+            # Move batch to the correct device
+            batch = accelerator.prepare_batch(batch)
+
+            batch_data, batch_targets = batch
```
Targets are just random values. 

In this test example - I created 10 feature X, and a 1 feature y (target). We use the 10 random features to predict the 1 target value. I added a quick node about the dataset as a comment in the __main__.

mlahariya commented on tests/ml_tools/test_accelerator.py:
```diff
@@ -0,0 +1,281 @@
+    assert prepared_optimizer is optimizer
+
+
+def test_prepare_data_without_distributed() -> None:
+    accelerator = Accelerator(spawn=False)
+    accelerator.world_size = 1
+    dl = get_dummy_dataloader()
+    prepared_dl = accelerator._prepare_data(dl)
+    assert prepared_dl is dl
+
+
+def test_prepare_data_with_distributed() -> None:
+    accelerator = Accelerator(spawn=False)
+    accelerator.world_size = 2
+    accelerator.local_rank = 0
+    dl = get_dummy_dataloader()
+    prepared_dl = accelerator._prepare_data(dl)
+    assert isinstance(prepared_dl.sampler, DistributedSampler)  # type: ignore[union-attr]
```
Updated.

RolandMacDoland reviewed (approved):
Thanks again @mlahariya. I guess I run out of comments from my side :) Minor ones so far, otherwise LGTM.

mlahariya commented on tests/ml_tools/test_accelerator.py:
```diff
@@ -0,0 +1,281 @@
+    accelerator.world_size = 1
+    accelerator.device = "cpu"
+    model = DummyModel()
+    prepared_model = accelerator._prepare_model(model)
```
Its not.  Only .prepare() is exposed to the user. 

This is just a test for the _prepare_model().

mlahariya commented on qadence/ml_tools/train_utils/strategy.py:
```diff
@@ -0,0 +1,350 @@
+        self.log_setup: str = log_setup
+        self.strategy: str
+        self.rank: int
+        self.world_size: int
```
Updated correctly.

mlahariya commented:
Alright everyone! @RolandMacDoland , @chMoussa , @debrevitatevitae.
This is ready. 

As suggested by @debrevitatevitae  - I have refactored the distribution strategy into two parts - `Execution` and `Distributor`. `Execution` - has an abstract base class, that can be inherited into required executions (with default, and torch run already available). It takes care of processes based env variables, as well as device dtype for each process. `Distributor` Takes care of distribution of processes, as well as cleaning up.

As suggested by @RolandMacDoland , I have removed the `spawn` variable. Only `nprocs` can be used to spawn multiple processes. 

There are few minor doc changes here and there, and all other review resolutions. Please have a look and let me know if you have more suggestions, otherwise I will merge it. 

The updated architecture. 
![image](https://github.com/user-attachments/assets/3c1d041f-c8e3-4fe6-baab-ac3a45a747f4)

debrevitatevitae commented:
LGTM too, thanks for addressing the comments!
